### PR TITLE
Add tasklist traffic metrics for non-sticky and non-forwarded tasklists

### PR DESCRIPTION
**Pull request by timl3136**
**What changed?**
Add a new emit metrics to track the rate of task being added into non-sticky and non-forwarded tasklists.

**Why?**
This metric allow us to better monitor tasklist and determine whether we need to partition tasklist if the tasklist becomes too long.

**How did you test it?**
Test both on existing local unit tests and on staging env.

**Potential risks**
Got blocked due to reaching series limit on emitted data.

**Release notes**

**Documentation Changes**

### CODE REVIEW

**Review by Shaddoll — COMMENTED** [review]
"This metric allow us to better monitor tasklist and determine whether we need to partition tasklist if the tasklist becomes too long."

If this is the goal, this metric is not necessary. `task_backlog_per_tl` and `task_lag_per_tl` is enough to tell us if a tasklist has too many tasks

File: service/matching/matchingEngine.go [hunk 1]
```diff
@@ -351,6 +366,21 @@ func (e *matchingEngineImpl) AddActivityTask(
 		return false, err
 	}
 
+	// get the domainName
+	domainName, err := e.domainCache.GetDomainName(domainID)
+	if err != nil {
+		return false, err
+	}
+
+	// tell if the tasklist is partitioned or sticky
```
Comment (neil-xie) [inline service/matching/matchingEngine.go:375]: could you add one more line of comment here to say the purpose so other people can know why you emit the metrics here, just like the summary you added for the pr

File: service/matching/matchingEngine.go [hunk 2]
```diff
@@ -296,6 +296,21 @@ func (e *matchingEngineImpl) AddDecisionTask(
...
+	if int32(request.GetTaskList().GetKind()) == 0 && request.ForwardedFrom == "" {
```
Comment (neil-xie) [inline service/matching/matchingEngine.go:306]: taskListKind was retrieved at the beginning of the method, you can just use it

**Comment by allenchen2244** [comment]
> "This metric allow us to better monitor tasklist and determine whether we need to partition tasklist if the tasklist becomes too long."
> 
> If this is the goal, this metric is not necessary. `task_backlog_per_tl` and `task_lag_per_tl` is enough to tell us if a tasklist has too many tasks

There's a few other things that we are gonna need this for. Such as creating a client dashboard that shows tasklist traffic percentage which I don't think we have.

File: service/matching/matchingEngine.go (hunk 1 above)
Comment (timl3136) [reply service/matching/matchingEngine.go]: Resolved in newest commit

File: service/matching/matchingEngine.go (hunk 2 above)
Comment (timl3136) [reply service/matching/matchingEngine.go]: Resolved in newest commit

File: service/matching/matchingEngine.go [hunk 3]
```diff
@@ -296,6 +296,21 @@ func (e *matchingEngineImpl) AddDecisionTask(
...
+	// get the domainName
+	domainName, err := e.domainCache.GetDomainName(domainID)
+	if err != nil {
+		return false, err
+	}
+
+	// Only emit traffic metrics if the tasklist is not sticky and is not forwarded
+	if taskListKind != nil && *taskListKind != types.TaskListKindSticky && request.ForwardedFrom == "" {
+		e.metricsClient.Scope(metrics.MatchingAddTaskScope).Tagged(metrics.DomainTag(domainName),
+			metrics.TaskListTag(taskListName), metrics.TaskListTypeTag("activity_task")).IncCounter(metrics.CadenceTasklistRequests)
```
Comment (neil-xie) [inline service/matching/matchingEngine.go:308]: Tasklist type tag was changed from "decision_task" to "activity_task" in your latest commit, is it by accident or you combined the types?

Comment (timl3136) [reply service/matching/matchingEngine.go]: It was an error, ty for spotting

**Review by allenchen2244 — APPROVED** [review]
I think we're good to go!

**Review by neil-xie — APPROVED** [review]

**Pull request merged**